### PR TITLE
action: run on windows

### PR DESCRIPTION
**Pull request by v1v**
This is one of the very last things to be migrated besides of the benchmarks or releases.

### Background

https://github.com/elastic/apm-agent-nodejs/pull/1393 was the original request that enabled Windows using docker linux containers for running all the required services.

### Blockers

Cannot run docker containers on windows. At present docker operations is not supported on Windows runners -> https://github.com/actions/runner/issues/904

### Questions

- ~~Is this particular stage needed?~~
- ~~If so, is it needed to run always on PRs? Or nightly could be a good solution?~~

### CODE REVIEW

**Comment by trentm** [comment]
> * Is this particular stage needed?

I would like to have *some* basic testing of the APM agent on Windows, yes.

However, I'd be fine if we skipped any of the test files that require an external service (like redis, postgres, etc.).
I think the way I would do that is:

```diff
diff --git a/test/instrumentation/modules/redis.test.js b/test/instrumentation/modules/redis.test.js
index 73d0962e..1fd3d8c8 100644
--- a/test/instrumentation/modules/redis.test.js
+++ b/test/instrumentation/modules/redis.test.js
@@ -9,6 +9,11 @@
 const redisVersion = require('redis/package.json').version
 const semver = require('semver')

+if (process.env.GITHUB_ACTIONS === 'true' && process.platform === 'win32') {
+  console.log('# SKIP: GH Actions do not support docker services on Windows')
+  process.exit(0)
+}
+
 if (semver.lt(redisVersion, '4.0.0')) {
   console.log('# SKIP: skipping redis.test.js tests <4.0.0')
   process.exit(0)
```

We'd need to do this to a lot of test files, but that should be okay. I can help with this if you like.
The easiest way to get the list of relevant test files to modify is by looking at what test files are run for each of these services/modules in ".tav.yml".


> * If so, is it needed to run always on PRs? Or nightly could be a good solution?

Does running it only nightly (and not on PRs) help with this issue? We still wouldn't be able to run the full test suite on a GitHub Actions windows worker.

**Comment by v1v** [comment]
Thanks @trentm , I think I can take it from here.

On the other hand, I'd like to extend the proposal to support external services on Windows if needed in the future with something like:

```js
if (process.env.SKIP_EXTERNAL_SERVICE_TEST === 'true' && process.platform === 'win32') {
```

Then we could support the below workflows:

1) Running on PRs for Windows with `SKIP_EXTERNAL_SERVICE_TEST=true`
2) Running on main for windows.

The `1)` will be done within this PR.

The `2)` will require running on private runners or the Cloud but that's something we could implement in a follow up, or at least think about how to support it. One of the reasons for splitting that logic in two different workflows is the one accessing to secrets, that's something we try to be cautious and avoid running on PRs with access to any secrets.

What do you think?

**Comment by trentm** [comment]
> On the other hand, I'd like to extend the proposal to support external services on Windows if needed in the future with something like:

Personally I would bias to *not* adding a `SKIP_EXTERNAL_SERVICE_TEST` abstraction unless we think it is likely that we will do `2) running on private runners or the Cloud`. However, I am not opposed.

If you choose to add the abstraction, perhaps a name like `TEST_NO_EXTERNAL_SERVICES`.
Also, shouldn't that envvar encode whether this is for Windows? I.e. it would be:

```
if (process.env.TEST_NO_EXTERNAL_SERVICES === 'true') { ...
```

**Comment by v1v** [comment]
There are some test failures on Windows:

The below one seems to be solved with https://github.com/elastic/apm-agent-nodejs/pull/3223/commits/c44ca29ace00bf0b92ddd0f63dfac989c4a30e57:

```
2023-03-30T08:20:25.7631353Z not ok 2 body should be uncompressed
2023-03-30T08:20:25.7631824Z   ---
2023-03-30T08:20:25.7632138Z     operator: equal
2023-03-30T08:20:25.7632507Z     expected: '/*\n * Copyri'
2023-03-30T08:20:25.7632874Z     actual:   '/*\r\n * Copyr'
2023-03-30T08:20:25.7633611Z     at: <anonymous> (D:\a\apm-agent-nodejs\apm-agent-nodejs\test\instrumentation\modules\http\github-423.test.js:57:11)
2023-03-30T08:20:25.7634081Z     stack: |-
2023-03-30T08:20:25.7634384Z       Error: body should be uncompressed
2023-03-30T08:20:25.7634970Z           at Test.assert [as _assert] (D:\a\apm-agent-nodejs\apm-agent-nodejs\node_modules\tape\lib\test.js:312:48)
2023-03-30T08:20:25.7635615Z           at Test.bound [as _assert] (D:\a\apm-agent-nodejs\apm-agent-nodejs\node_modules\tape\lib\test.js:95:17)
2023-03-30T08:20:25.7636238Z           at Test.strictEqual (D:\a\apm-agent-nodejs\apm-agent-nodejs\node_modules\tape\lib\test.js:476:7)
2023-03-30T08:20:25.7636882Z           at Test.bound [as strictEqual] (D:\a\apm-agent-nodejs\apm-agent-nodejs\node_modules\tape\lib\test.js:95:17)
2023-03-30T08:20:25.7637522Z           at D:\a\apm-agent-nodejs\apm-agent-nodejs\test\instrumentation\modules\http\github-423.test.js:57:11
2023-03-30T08:20:25.7638031Z           at processTicksAndRejections (internal/process/task_queues.js:95:5)
2023-03-30T08:20:25.7638354Z   ...
```

but there is now another one:

```
2023-03-30T08:47:15.6808149Z running test: cd . && node --unhandled-rejections=strict test\opentelemetry-bridge\fixtures.test.js
2023-03-30T08:47:15.9155107Z TAP version 13
...
2023-03-30T08:47:19.4083062Z D:\a\apm-agent-nodejs\apm-agent-nodejs\test\opentelemetry-bridge\fixtures.test.js:120
2023-03-30T08:47:19.4083778Z       const portB = tas[3].span.context.destination.port
2023-03-30T08:47:19.4084476Z                                 ^
2023-03-30T08:47:19.4084751Z 
2023-03-30T08:47:19.4085649Z TypeError: Cannot read property 'context' of undefined
2023-03-30T08:47:19.4086799Z     at Object.check (D:\a\apm-agent-nodejs\apm-agent-nodejs\test\opentelemetry-bridge\fixtures.test.js:120:33)
2023-03-30T08:47:19.4087779Z     at done (D:\a\apm-agent-nodejs\apm-agent-nodejs\test\opentelemetry-bridge\fixtures.test.js:427:15)
2023-03-30T08:47:19.4088457Z     at ChildProcess.exithandler (child_process.js:374:7)
2023-03-30T08:47:19.4089040Z     at ChildProcess.emit (events.js:400:28)
2023-03-30T08:47:19.4089549Z     at maybeClose (internal/child_process.js:1088:16)
2023-03-30T08:47:19.4090154Z     at Process.ChildProcess._handle.onexit (internal/child_process.js:296:5)
2023-03-30T08:47:19.4216014Z D:\a\apm-agent-nodejs\apm-agent-nodejs\test\test.js:218
2023-03-30T08:47:19.4216534Z     if (err) throw err
2023-03-30T08:47:19.4216823Z              ^
2023-03-30T08:47:19.4217200Z 
2023-03-30T08:47:19.4217788Z Error: non-zero error code
2023-03-30T08:47:19.4221637Z     at ChildProcess.<anonymous> (D:\a\apm-agent-nodejs\apm-agent-nodejs\test\test.js:106:21)
2023-03-30T08:47:19.4222783Z     at ChildProcess.emit (events.js:400:28)
2023-03-30T08:47:19.4223307Z     at maybeClose (internal/child_process.js:1088:16)
2023-03-30T08:47:19.4223920Z     at Process.ChildProcess._handle.onexit (internal/child_process.js:296:5) {
2023-03-30T08:47:19.4224487Z   code: 'ENONZERO',
2023-03-30T08:47:19.4225110Z   exitCode: 1
2023-03-30T08:47:19.4225496Z }
```

any hints?

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -19,11 +19,10 @@ pipeline {
     quietPeriod(10)
   }
   triggers {
-    issueCommentTrigger("(${obltGitHubComments()}|^run (module|benchmark) tests.*)")
+    issueCommentTrigger("(^run benchmark tests.*)")
```
Comment (v1v) [inline .ci/Jenkinsfile:22]:
```suggestion
    issueCommentTrigger("^run benchmark tests.*")
```

**Comment by trentm** [comment]
@v1v Instead of the `.gitattributes` change in  https://github.com/elastic/apm-agent-nodejs/commit/c44ca29ace00bf0b92ddd0f63dfac989c4a30e57 can we please make this change:

```diff
diff --git a/test/instrumentation/modules/http/github-423.test.js b/test/instrumentation/modules/http/github-423.test.js
index a00bebfb..e74e2fe5 100644
--- a/test/instrumentation/modules/http/github-423.test.js
+++ b/test/instrumentation/modules/http/github-423.test.js
@@ -54,7 +54,7 @@ test('https://github.com/elastic/apm-agent-nodejs/issues/423', function (t) {
     var server = http.createServer(function (req, res) {
       got(url).then(function (response) {
         t.strictEqual(response.body.length, fileSize, 'body should be expected size')
-        t.strictEqual(response.body.slice(0, 12), '/*\n * Copyri', 'body should be uncompressed')
+        t.ok(response.body.includes('Copyright Elasticsearch'), 'body should be uncompressed')
         res.end()
       })
     })
```

**Comment by trentm** [comment]
> @v1v Instead of the `.gitattributes` change in [c44ca29](https://github.com/elastic/apm-agent-nodejs/commit/c44ca29ace00bf0b92ddd0f63dfac989c4a30e57) can we please make this change:

I've done it.

I'll try to fix the "test\opentelemetry-bridge\fixtures.test.js" failure as well.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -1,12 +1,4 @@
 # Run a significant subset of tests on GitHub Actions.
-#
-# Examples of things *not* being run here currently that are run by
-# the Jenkins-based tests:
-# - a test run on windows
-# - the "TAV" tests that run against all supported versions of instrumented
-#   modules
-# - the integration tests that run with an apm-server
-# - code coverage (npm run coverage)
```
Comment (trentm) [inline .github/workflows/test.yml:2]: Can drop that first line too. It is no longer a subset of tests.

**Review by trentm — APPROVED** [review]

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -161,3 +150,15 @@ jobs:
...
         type: "junit"
         githubToken: ${{ secrets.GITHUB_TOKEN }}
+
+  test-windows:
+    runs-on: windows-latest
+    steps:
+    - uses: actions/checkout@v3
+    - uses: actions/setup-node@v3
+      with:
+        node-version: 14
```
Comment (kuisathaverat) [inline .github/workflows/test.yml:160]: lest make it faster 😄 
```suggestion
        node-version: 14
        cache: 'npm'
```

**Review by kuisathaverat — APPROVED** [review]

**Pull request merged**